### PR TITLE
Create new engine module 

### LINKED ISSUE #2449: Logstream and Broker-core highly depend on each other
### Problem

Currently we have the problem that the modules (logstream and broker-core) are splitted, but highly depend on each other (dependency problem).

This means for example the StreamProcessorController don't know explicit stuff, like events or state types (ColumnFamilies etc.), but it needs some information's to work as expected. For these information we added some abstractions, suppliers or high order functions etc.

##### Blacklisting:
  * introduces an new callback **onError**
  * controller depends on the implementation
    * that workflow instance is added to state
    * that an error event is written (events are not materialized in the logstream)
    * controller waits until last written position is committed
  ##### Async snapshotting:
   * introduces new method position to recovery from
     * **before**: exporter uses reader of controller and seeks to the position he needs (really hacky!)
       * gives now position to recovery from
     * processor stores position of last processed event and gives this on recovery back
       * necessary because of the async snapshotting
       * might be better that the controller stores the position in the DB after the processing was successful and restores this position

### Solution:

Create an engine module which contains the core components of zeebe, such that we have not weird module relations/dependencies and hacky solutions to solve that.

#### Idea: 

Classes/Packages which should probably be in the same module:

 * io.zeebe.logstreams.processor.*
* io.zeebe.logstreams.state.*
* io.zeebe.broker.logstreams.*
* io.zeebe.broker.workflow.*
* io.zeebe.broker.job.*
* io.zeebe.broker.incident.*

*Note:* We should think about licensing

### PR DESCRIPTION
**Pull request by ChrisKujawa**
* a new module called engine was created to solve the dependency problems (coupling) between the logstreams and broker-core module 
    * this module contains all stream processing (+ state management)
 * most of the tests are migrated to new module
    * there were some test which can't be moved yet https://github.com/zeebe-io/zeebe/issues/2461
 * snapshot controller needs to be moved later https://github.com/zeebe-io/zeebe/issues/2460
 * introduces new interfaces to avoid coupling (interfaces for subscription command sender etc.)

closes #2449

### CODE REVIEW

File: broker-core/src/main/java/io/zeebe/broker/engine/EngineServiceNames.java [hunk 1]
```diff
@@ -32,6 +32,11 @@
         String.format("%s.rocksdb.storage", partitionName), StateStorageFactory.class);
   }
 
-  public static final ServiceName<ZbStreamProcessorService> ZB_STREAM_PROCESSOR_SERVICE_NAME =
-      ServiceName.newServiceName("logstreams.processor", ZbStreamProcessorService.class);
+  public static final ServiceName<EngineService> ZB_STREAM_PROCESSOR_SERVICE_NAME =
```
Comment (MiguelPires) [inline broker-core/src/main/java/io/zeebe/broker/engine/EngineServiceNames.java]: Maybe we can rename this to ENGINE_SERVICE_NAME so it's consistent with the rest. What do you think?

**Review by miguelpires — CHANGES_REQUESTED** [review]
Looks good. Just pointed out some minor things

**Comment by Zelldon** [comment]
Thanks @MiguelPires will change it accordingly

**Comment by Zelldon** [comment]
@MiguelPires please see the lates changes

**Review by miguelpires — APPROVED** [review]

**Pull request merged**